src/hotspot/share/gc/g1/g1CollectedHeap.cpp

```diff
@@ -989,82 +946,153 @@ HeapWord* G1CollectedHeap::satisfy_failed_allocation_helper(size_t word_size,
+bool G1CollectedHeap::satisfy_failed_allocations(bool* gc_succeeded) {
   assert_at_safepoint_on_vm_thread();
 
-  // Attempts to allocate followed by Full GC.
-  HeapWord* result =
-    satisfy_failed_allocation_helper(word_size,
-                                     true,  /* do_gc */
-                                     false, /* maximum_collection */
-                                     false, /* expect_null_mutator_alloc_region */
-                                     succeeded);
+  bool success = handle_allocation_requests(false /* expect_null_mutator_alloc_region*/);
 
-  if (result != nullptr || !*succeeded) {
-    return result;
+  if (success) {
+   return success;
   }
 
-  // Attempts to allocate followed by Full GC that will collect all soft references.
-  result = satisfy_failed_allocation_helper(word_size,
-                                            true, /* do_gc */
-                                            true, /* maximum_collection */
-                                            true, /* expect_null_mutator_alloc_region */
-                                            succeeded);
+  // Attempt to satisfy allocation requests failed; reset the requests, execute a full-gc,
+  // then try again
+  reset_allocation_requests();
 
-  if (result != nullptr || !*succeeded) {
-    return result;
+  *gc_succeeded = do_full_collection(false /* clear_all_soft_refs */, false /* do_maximal_compaction */);
+
+  if (!*gc_succeeded) {
+    return false;
   }
 
-  // Attempts to allocate, no GC
-  result = satisfy_failed_allocation_helper(word_size,
-                                            false, /* do_gc */
-                                            false, /* maximum_collection */
-                                            true,  /* expect_null_mutator_alloc_region */
-                                            succeeded);
+  success = handle_allocation_requests(true /* expect_null_mutator_alloc_region*/);
 
-  if (result != nullptr) {
-    return result;
+  if (success) {
+   return success;
+  }
+
+  // Attempt to satisfy allocation requests after full-gc also failed. We reset the allocation requests
+  // then execute a maximal compaction full-gc before retrying the allocations
+
+  reset_allocation_requests();
+
+  *gc_succeeded = do_full_collection(true /* clear_all_soft_refs */, true /* do_maximal_compaction */);;
+
+  if (!*gc_succeeded) {
+    return false;
+  }
+
+  success = handle_allocation_requests(true /* expect_null_mutator_alloc_region*/);
+
+  if (success) {
+   return success;
+  }
+
+  // Even after maximal compaction full gc, we failed to satisfy all the pending allocation requests.
+  // We can declare the pending requests as failed.
+  DoublyLinkedList<StalledAllocReq>::RemoveIterator iter(&_stalled_allocations);
+
+  for (StalledAllocReq* alloc_req; iter.next(&alloc_req);) {
+    _satisfied_allocations.insert_last(alloc_req);
+    alloc_req->set_state(StalledAllocReq::AllocationState::Failed);
```

Comment:
Maybe that's just me but I prefer code changing the request first and then inserting it. I.e. process then pass on to something else instead of the other way around.